src/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -453,37 +465,73 @@ private int InternalValueCountCore()
         /// <returns>All value names.</returns>
         private unsafe string[] GetValueNamesCore(int values)
         {
-            string[] names = new string[values];
-            char[] name = new char[MaxValueLength + 1];
-            int namelen;
+            List<string> names = new List<string>(values);
+
+            // Names in the registry aren't usually very long, although they can go to as large
+            // as 16383 characters (MaxValueLength).
+            //
+            // Every call to RegEnumValue will allocate another buffer to get the data from
+            // NtEnumerateValueKey before copying it back out to our passed in buffer. This can
+            // add up quickly- we'll try to keep the memory pressure low and grow the buffer
+            // only if needed.
 
-            fixed (char* namePtr = &name[0])
+            char[] name = ArrayPool<char>.Shared.Rent(100);
+
+            try
             {
-                for (int i = 0; i < values; i++)
+                int index = 0;
+                int result;
+                int nameLength = name.Length;
+
+                while ((result = Interop.Advapi32.RegEnumValue(
+                    _hkey,
+                    index,
+                    name,
+                    ref nameLength,
+                    IntPtr.Zero,
+                    null,
+                    null,
+                    null)) != Interop.Errors.ERROR_NO_MORE_ITEMS)
                 {
-                    namelen = name.Length;
-
-                    int ret = Interop.Advapi32.RegEnumValue(_hkey,
-                        i,
-                        namePtr,
-                        ref namelen,
-                        IntPtr.Zero,
-                        null,
-                        null,
-                        null);
-
-                    if (ret != 0)
+                    switch (result)
                     {
-                        // ignore ERROR_MORE_DATA if we're querying HKEY_PERFORMANCE_DATA
-                        if (!(IsPerfDataKey() && ret == Interop.Errors.ERROR_MORE_DATA))
-                            Win32Error(ret, null);
+                        // The size is only ever reported back correctly in the case
+                        // of ERROR_SUCCESS. It will almost always be changed, however.
+                        case Interop.Errors.ERROR_SUCCESS:
+                            names.Add(new string(name, 0, nameLength));
+                            index++;
+                            break;
+                        case Interop.Errors.ERROR_MORE_DATA:
+                            if (IsPerfDataKey())
+                            {
+                                // Enumerating the values for Perf keys always returns ERROR_MORE_DATA.
+                                names.Add(new string(name));
```

Comment:
Yeah, it makes no sense, but that is the way the API works with the perf key. It always returns an error with garbage in the length, but an actual value (null terminated). It isn't something you are really supposed to be doing with a perf key, but we supported it historically.